caip-browser_extension_provider_communication.md

```diff
@@ -0,0 +1,141 @@
+
+A standard method for providers to communicate with web extebsions would allow website authors to embed their own providers without losing web extension inter-operability.
+
+## Specification
+<!--The technical specification should describe the standard in detail. The specification should be detailed enough to allow competing, interoperable implementations. -->
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
+uppercase in this document are to be interpreted as described in [RFC
+2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Summary
+
+Web extensions should expose a standard interface over [`externally_connectable`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable) to enable the inter-operability of embedded providers.
+
+### Message format
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "/caip-x/schemas/provider-request.schema.json",
+  "title": "Provider / Web Extension message",
+  "description": "A request sent between a provider and a web extension.",
+  "type": "object",
+  "properties": {
+    "type": {
+      "description": "The message type, used to identify this as a CAIP-X message",
+      "const": "caip-x"
+    },
+    "data": {
+      "description": "A JSON-RPC message",
+      "type": "object"
+    }
+  },
+  "required": ["type", "data"]
+}
+```
+
+TODO: should type be more defined here? Seems weird to specify the envelop but no types it can be used with?
+Does `data` need to be a JSON-RPC message specifically?
+Should `type` be something besides `caip-x`?
+
+
+### `externally_connectable`
+
+A web extension can use the `externally_connectable` manifest field to accept messages from websites and other extensions. This permission can be configured to limit which sites and other extensions can send messages to the web extension. How this permission is configured is out-of-scope for this proposal; this proposal only concerns sites that the web extension allows messages from.
+
+
+The web extension can:
+* handle a connection from the website or other extensions by using `chrome.runtime.onConnectExternal.addListener((port) => {...})`
+* send messages by using `port.postMessage()`
+* receive messages by using `port.onMessage.addListener()`
+  * incoming messages should be validated according to the message format specification above
+
+The provider can:
+* initiate a connection with the web extension by using `port = browser.runtime.connect()`
+* send messages by using `port.postMessage()`
+* receive messages by using `port.onMessage.addListener()`
+
+Caveats:
+* No firefox support yet, but they are considering implementing it
+* Extension Ids may be different across different family of browsers
+  * Discoverability should address this
+
```

Comment:
not sure where to put these quite yet.

Also need to flesh out Extension IDs being different? Or do we think we can push that into the discoverability CAIP?

Reply:
Still unsure where this should live

Reply:
> Also need to flesh out Extension IDs being different? Or do we think we can push that into the discoverability CAIP?

yeah atleast on this point I can bring this into [the WIP discovery CAIP](https://github.com/adonesky1/CAIPs/pull/1)

Reply:
> No firefox support yet, but they are considering implementing it

This seems like the right place to mention this. Maybe we can find the right link to point to their indications of eventually supporting it

Reply:
both points addressed here https://github.com/jiexi/CAIPs/pull/1/commits/e7a8390561715af4f3cb8e6705a9fc94bf8d8ce4

leaving the different extension IDs to the discovery CAIP